docs/design/metadata.md

```diff
@@ -1,259 +1,284 @@
-# Metadata formats
+# Metadata format
 
 ## Purpose
 
-This document describes the three metadata formats used for variant
-wheels: `pyproject.toml` tables, Core Metadata fields (used
-in `*.dist-info/METADATA` and `PKG-INFO` files)
-and `variants.json` file. All three formats are related, as they share
-a common part and form a single pipeline:
+This document describes the metadata format used for variant wheels.
+The format is used in three locations, with slight variations:
 
-    pyproject.toml → METADATA in wheels → variants.json
+1. in the source repository, as a `pyproject.toml` file,
 
-The `pyproject.toml` file is used to build the wheels. As a common
-project configuration format, it also specifies the metadata used later
-to install them.
+2. in the built wheel, as a `*.dist-info/variant.json` file,
 
-When the wheels are built, the build backend transfers the relevant
-metadata into the `METADATA` file, along with the built variant
-description. This information is used when installing the wheels.
+3. on the wheel index, as a `*-variants.json` file.
 
-Additionally, for the purpose of indexing, the metadata from multiple
-wheels is collected into a single `variants.json` format file. This file
-can be afterwards fetched by the installers to determine which variants
-are available and how to install the provider plugins required to
-determine which variants are supported. This makes it possible to select
-a specific variant wheel without having to download them all.
+All three variants share a common structure, with some of its element
+shared across all of them, and some being specific to a single variant,
+as described further in the document. This structure is then serialized
+into TOML or JSON appropriately.
 
+These variations fit into the common wheel building pipeline where
+a source tree is used to build one or more wheels, and the wheels
+are afterwards published on an index. The `pyproject.toml` file
+provides the metadata needed to build the wheels, as well as the shared
+metadata needed to install them. This metadata is then amended with
+information on the specific variant build, and copied into the wheel.
+When wheels are uploaded into the index, the metadata from all of them
+is read and merged into a single JSON file that can be used
```

Comment:
Tangential to this PR but: there should be a way/tool to create the `*-variants.json` file locally on a set of built wheels. Not everyone runs a full index with triggers on upload that can run arbitrary tooling. Is there such a tool already? If so, let's refer to it. If not, let's discuss where it should live (e.g., standalone or as part of `variantlib` or an existing library).

Reply:
@rgommers same ! It took me a few days to appreciate this change to its value. The amount of simplification and unification it brings to the whole Variant concept is greatly appreciated ;)